gobblin-core/src/main/java/gobblin/source/extractor/extract/QueryBasedSource.java

```diff
@@ -12,236 +12,245 @@
-
+import com.google.common.collect.Sets;
+
+import gobblin.config.client.ConfigClient;
+import gobblin.config.client.ConfigClientCache;
+import gobblin.config.client.api.ConfigStoreFactoryDoesNotExistsException;
+import gobblin.config.client.api.VersionStabilityPolicy;
+import gobblin.config.store.api.ConfigStoreCreationException;
+import gobblin.config.store.api.VersionDoesNotExistException;
 import gobblin.configuration.ConfigurationKeys;
 import gobblin.configuration.SourceState;
+import gobblin.configuration.State;
 import gobblin.configuration.WorkUnitState;
 import gobblin.configuration.WorkUnitState.WorkingState;
 import gobblin.source.extractor.JobCommitPolicy;
+import gobblin.source.extractor.WatermarkInterval;
 import gobblin.source.extractor.partition.Partitioner;
 import gobblin.source.extractor.utils.Utils;
-import gobblin.source.extractor.watermark.WatermarkPredicate;
-import gobblin.source.extractor.watermark.WatermarkType;
 import gobblin.source.workunit.Extract;
 import gobblin.source.workunit.Extract.TableType;
+import gobblin.source.workunit.MultiWorkUnit;
 import gobblin.source.workunit.WorkUnit;
+import gobblin.util.ConfigUtils;
+import gobblin.util.PathUtils;
+import gobblin.util.dataset.DatasetUtils;
+import lombok.extern.slf4j.Slf4j;
 
 
 /**
  * A base implementation of {@link gobblin.source.Source} for
  * query-based sources.
  */
+@Slf4j
 public abstract class QueryBasedSource<S, D> extends AbstractSource<S, D> {
 
-  private static final Logger LOG = LoggerFactory.getLogger(QueryBasedSource.class);
+  private static final String QUERY_BASED_SOURCE = "query_based_source";
 
   @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
     initLogger(state);
 
     List<WorkUnit> workUnits = Lists.newArrayList();
     String nameSpaceName = state.getProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY);
-    String entityName = state.getProp(ConfigurationKeys.SOURCE_ENTITY);
 
-    String extractTableName = state.getProp(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY);
-    // If extract table name is not found then use the entity name
-    if (StringUtils.isBlank(extractTableName)) {
-      extractTableName = Utils.escapeSpecialCharacters(entityName, ConfigurationKeys.ESCAPE_CHARS_IN_TABLE_NAME, "_");
+    Map<String, String> tableNameToEntityMap = Maps.newHashMap();
+    for (String entity : getSourceEntities(state)) {
+      tableNameToEntityMap.put(Utils.escapeSpecialCharacters(entity, ConfigurationKeys.ESCAPE_CHARS_IN_TABLE_NAME, "_"),
+          entity);
     }
 
-    TableType tableType = TableType.valueOf(state.getProp(ConfigurationKeys.EXTRACT_TABLE_TYPE_KEY).toUpperCase());
-    long previousWatermark = this.getLatestWatermarkFromMetadata(state);
-
-    Map<Long, Long> sortedPartitions = Maps.newTreeMap();
-    sortedPartitions.putAll(new Partitioner(state).getPartitions(previousWatermark));
+    Map<String, State> tableSpecificPropsMap = shouldObtainTablePropsFromConfigStore(state)
```

Comment:
This looks dangerous. The config store may be needed for some completely different purpose (e.g. in a converter or writer). With this change, the source is forced to support the config store. We want to go there eventually, but we are not there yet.